Running OK on Firefox

orlandoaleman opened the pull request:
Hello there,
firstly congratulate you on the project. It is very well coded and organized. It is highly readable.

I am sending you a pull request with some corrections that I have made to make it work in Firefox and there is (in my opinion) more consistency.

Thanks for sharing.

Greetings

orlandolaycos committed:
ScarletsAudioStreamer, ScarletsMediaPresenter:
- onStop callback for more granulated flow control

ScarletsMediaPresenter:
- reset the buffer header on every play
- check if passed mimeType is a supported mimeType
- MediaRecorder only supports 1 parameters. this throw error on firefox

FIX latency NaN in realtimeBufferPlay and refactor local variables for coherence

I've fixed a bad calc of the latency that causes NaN values at the method 'realtimeBufferPlay' . Also I've refactored some local variables for better coherence

Refactor local variables for coherence

I've refactored some local variables for better coherence

FIXED header latency when is a video

More fixes. Running on chrome, firefox, safari

orlandolaycos committed:
FIX stopRecording being called after denying micro

orlandolaycos committed:
dist/ updated

orlandolaycos committed:
MediaBuffer available as module

orlandolaycos committed:
dist/ updated

StefansArya commented:
Hi! Welcome to open source library!
I really appreciate your compliment and also the information you provided on the changes.

By the way, looks like you're having trouble when using the library for Firefox before..
Have you tried using the polyfill instead? https://github.com/ScarletsFiction/SFMediaStream/issues/15#issuecomment-757400133

Usually the polyfill could help some browser like Firefox for recording audio using Opus encoder.

I will merge the pull request after reviewing and testing the changes when I have time.
Anyway, thank you for your contribution!

StefansArya commented on src/AudioStreamer.js:
```diff
@@ -124,7 +125,7 @@ var ScarletsAudioStreamer = function(chunksDuration){
 
 		else // Direct output to destination
-			scope.bufferElement[index].connect(scope.audioContext.destination);
+			scope.bufferElement[index].connect(scope.audioContext.destinatgion);
```
I think it should be `.destination` instead of `.destinatgion` 😄 
It's connecting the [AudioBuffer node](https://developer.mozilla.org/en-US/docs/Web/API/BaseAudioContext/createBufferSource) to [AudioContext](https://developer.mozilla.org/en-US/docs/Web/API/AudioContext#examples)'s destination if the stream is not using Plugin/Effect

orlandoaleman replied:
What a shame! Sorry

StefansArya commented on src/MediaPresenter.js:
```diff
@@ -97,18 +102,19 @@ var ScarletsMediaPresenter = function(options, latency){
 		var bufferHeaderLength = false;
 
-		scope.mediaRecorder = new MediaRecorder(mediaStream, options, scope.polyfill);
+		scope.mediaRecorder = new MediaRecorder(mediaStream, options);
```
I think we will need to keep the `scope.polyfill` on there, because when we're using the [recorder polyfill](https://github.com/kbumsik/opus-media-recorder#working-with-a-bundler) someone may want to add custom option for modifying the `workerOptions` 🤔

orlandoaleman replied:
Would it be possible to detect if the native MediaRecorder class or a polyfill? 
On my setup (latest Firefox Dev Edition, Fedora Linux 34) this third param throws error.

I'll try to test this on Monday, when I get back to the office.

StefansArya replied:
Hmm, looks like you're right.. it's also throw error when the MediaRecorder was given the third param even it was `undefined` value on Firefox 90, while chromium based browser works normally.

By the way, the polyfill docs said if the `encoderWorker.js` already loaded using the `<script>` tag, we can leave `workerOptions` or `scope.polyfill` empty..

Let's use your implementation 👍

orlandoaleman replied:
what do you about something similar this?

```
  let usingOpusMediaRecorderPolyfill = false;
  if (window.OpusMediaRecorder) {
	  if (options.alwaysUsePolyfill) {
		  usingOpusMediaRecorderPolyfill = true;
	  }
	  else if (!window.MediaRecorder) {
		  usingOpusMediaRecorderPolyfill = true;
	  }
	  else {
		  usingOpusMediaRecorderPolyfill = window.MediaRecorder === window.OpusMediaRecorder;
	  }
  }
  
  if (usingOpusMediaRecorderPolyfill) {
	  scope.mediaRecorder = new OpusMediaRecorder(mediaStream, options, scope.polyfill);
  }
  else {
	  scope.mediaRecorder = new MediaRecorder(mediaStream, options);
  }
```

orlandoaleman replied:
After thinking about it for a while, I think that removing it is the correct option. Although I will send you another pull request with a proposal based on [the previous comment](https://github.com/ScarletsFiction/SFMediaStream/pull/21#discussion_r697913484)). Feel free to accept or reject it at your discretion 😇 😉

StefansArya replied:
Yes, I already thought similar method like your comment 😄, but I haven't seen someone posted an issue about the error when the `MediaRecorder` is given the third parameter. Maybe they mostly using chromium based browser or already using the polyfill with the `encoderWorker.js`.. so they wasn't seen that error 🤔 

I think the current pull request is already OK, Thanks! 😉

orlandoaleman committed:
FIX typo

StefansArya commented on src/z_extra.js:
```diff
@@ -169,6 +170,7 @@ else{
 	global.ScarletsVideoStreamer = ScarletsVideoStreamer;
 	global.ScarletsMediaPlayer = ScarletsMediaPlayer;
 	global.ScarletsMediaPresenter = ScarletsMediaPresenter;
+	global.MediaBuffer = MediaBuffer;
```
Can you change this line to `global.ScarletsMediaBuffer = MediaBuffer;` ?

because it would add `MediaBuffer` as a global variable in `window.MediaBuffer`, we will need to add a prefix to avoid the `MediaBuffer` name conflicting with other variable on the developer/user side.

Commit:
Changed usage of MediaBuffer as global renaming to ScarletsMediaBuffer
Explained here:
https://github.com/ScarletsFiction/SFMediaStream/pull/21#discussion_r697904440

dist/ updated

Commit:
FIX

Commit:
dist/
